File: src/main/kotlin/m7_wrapup/Streams.java
package m7_wrapup;

import java.util.Arrays;
import java.util.Objects;
import java.util.stream.Collectors;

public class Streams {

    public static void main(String[] args) {

        var list = Arrays.asList("Java", "kotlin", "C", "C++", "c#", null);

        list.stream()
                .filter(Objects::nonNull)
                .map(String::toUpperCase)
                .collect(Collectors.toMap(str -> str, String::length))
                .forEach((k,v) -> System.out.println(k + "=" + v));
    }
}

File: src/main/kotlin/m7_wrapup/Streams.kt
package for_java_devs.m7wrapup.collections


fun main() {
    val languages = listOf("JavaScript", "Kotlin", "java", "C", "C++", "c#", null)


    languages
        .filterNotNull()
        .filter { !it.startsWith("C") }
        .map { it.uppercase() }
        .associateWith { it.length }
        .forEach { println(it) }


    val list = languages.asSequence()
        .filterNotNull()
        .filter {
            println("Filtering $it")
            it.length > 2
        }
        .map {
            println("Mapping $it")
            it.uppercase()
        }
        .take(2)

    println(list.joinToString())
}